### PR TITLE
Support for tsconfig.json

### PR DESCRIPTION
**Pull request by nycdotnet**
This PR adds `tsconfig.json` support and represents a significant refactoring of grunt-ts.

Documentation is here: https://github.com/TypeStrong/grunt-ts/issues/202

To Do:
- [x] Refactor parameters from task and target
- [x] Refactor `vs` implementation
- [x] Add fast unit testing support for gruntfile, `vs`, and tsconfig.json
- [x] Implement tsconfig.json support for `compilerOptions`
- [x] Implement tsconfig.json support for `files`
- [x] Implement tsconfig.json support for `exclude` array.
- [x] ignoreFiles
- [x] templates working for vs, tsconfig, and other parameters
- [x] warn on incorrect positioning of grunt-ts related configuration items in gruntfile
- [x] fix simple capitalization issues on grunt-ts related configuration items in gruntfile
- [x] implement ignoreSettings
- [x] implement overwriteFilesGlob
- [x] implement updateFiles
- [x] implement passThrough
- [x] fix warnings that were removed (look through integration tests in the gruntfile and re-implement using the new optionsResolver mechanism).
- [x] update documentation.

### CODE REVIEW

File: tasks/ts.js [hunk 1]
```diff
@@ -566,57 +277,34 @@ function pluginFn(grunt) {
...
-                        if (filesToCompile.length > 0) {
-                            return runCompilation(filesToCompile, rawTargetConfig, options).then(function (success) {
+                        if (filesToCompile.length > 0 || options.testExecute) {
+                            return runCompilation(options, currentFiles).then(function (success) {
                                 return success;
                             });
                         }
                         else {
+                            // Nothing to do
                             grunt.log.writeln('No files to compile'.red);
```
Comment (Eyas) [inline tasks/ts.js]: this error will trigger in the tsconfig passThrough case. Perhaps another `||` clause for that case?

File: tasks/ts.ts [hunk 2]
```diff
@@ -664,22 +339,24 @@ function pluginFn(grunt: IGrunt) {
...
-                            return runCompilation(filesToCompile, rawTargetConfig, options).then((success: boolean) => {
+                        if (filesToCompile.length > 0 || options.testExecute) {
+                            return runCompilation(options, currentFiles).then((success: boolean) => {
                                 return success;
                             });
                         }
-                        else { // Nothing to do
+                        else {
+                            // Nothing to do
                             grunt.log.writeln('No files to compile'.red);
```
Comment (Eyas) [inline tasks/ts.ts]: Same: this error will trigger in the tsconfig passThrough case. Perhaps another `||` clause for that case?

File: tasks/modules/compile.js [hunk 3]
```diff
@@ -63,33 +64,34 @@ function getTsc(binPath) {
...
                 files = newFiles;
                 // If outDir is specified but no baseDir is specified we need to determine one
-                if (target.outDir && !target.baseDir) {
-                    target.baseDir = utils.findCommonPath(files, '/');
+                if (compilationInfo.outDir && !options.baseDir) {
+                    options.baseDir = utils.findCommonPath(files, '/');
                 }
             }
             else {
                 exports.grunt.log.writeln('No file changes were detected. Skipping Compile'.green);
```
Comment (Eyas) [inline tasks/modules/compile.js]: Same: this error will trigger in the tsconfig passThrough case. Perhaps another `||` clause for that case?

**Comment by Eyas** [comment]
@nycdotnet I left a few comments on bugs with the `passThrough` case for tsconfig

**Comment by nycdotnet** [comment]
Thanks.  Will address this soon.

**Comment by nycdotnet** [comment]
Eyas - I released `grunt-ts@5.0.0-beta.6` which addresses these comments.  Would you please see if this resolves your issue?  Thanks.

**Pull request merged**